File: java_package/brainflow/src/main/java/brainflow/examples/ICA.java
package brainflow.examples;

import java.util.List;

import brainflow.BoardIds;
import brainflow.BoardShim;
import brainflow.BrainFlowInputParams;
import brainflow.DataFilter;

public class ICA
{

    public static void main (String[] args) throws Exception
    {
        // use synthetic board for demo
        BoardShim.enable_board_logger ();
        BrainFlowInputParams params = new BrainFlowInputParams ();
        BoardIds board_id = BoardIds.SYNTHETIC_BOARD;
        BoardShim board_shim = new BoardShim (board_id, params);
        board_shim.prepare_session ();
        board_shim.start_stream (3600);
        Thread.sleep (10000);
        board_shim.stop_stream ();
        double[][] data = board_shim.get_board_data (500);
        board_shim.release_session ();

        int[] eeg_channels = BoardShim.get_eeg_channels (board_id);
        int eeg_channel = eeg_channels[1];
        double[][] ica_data = DataFilter.reshape_data_to_2d (5, 100, data[eeg_channel]);
        List<double[][]> ica = DataFilter.perform_ica (ica_data, 2);
        System.out.println ("Completed");
    }
}

File: nodejs_package/tests/ica.ts
import {BoardIds, BoardShim, DataFilter} from 'brainflow';

function sleep (ms: number)
{
    return new Promise ((resolve) => { setTimeout (resolve, ms); });
}

async function runExample (): Promise<void>
{
    const boardId = BoardIds.SYNTHETIC_BOARD;
    const board = new BoardShim (boardId, {});
    board.prepareSession();
    board.startStream();
    await sleep (10000);
    board.stopStream();
    const data = board.getCurrentBoardData(500);
    board.releaseSession()
    const eegChannels = BoardShim.getEegChannels(boardId);
    const eegData = data[eegChannels[0]];
    const eeg2D: number[][] = [];
    while (eegData.length)
        eeg2D.push(eegData.splice(0, 100));
    const icaData = DataFilter.performIca(eeg2D, 2, [0, 1, 2, 3, 4]);
    console.info(icaData[3]);
}

runExample ();
